include/ginkgo/core/solver/idr.hpp

```diff
@@ -0,0 +1,222 @@
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_SOLVER_IDR_HPP_
+#define GKO_CORE_SOLVER_IDR_HPP_
+
+
+#include <random>
+#include <vector>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/log/logger.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/identity.hpp>
+#include <ginkgo/core/stop/combined.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+namespace gko {
+/**
+ * @brief The ginkgo Solve namespace.
+ *
+ * @ingroup solvers
+ */
+namespace solver {
+
+
+/**
+ * IDR(s) is an efficient method for solving large nonsymmetric systems of
+ * linear equations.
+ *
+ * The method is based on the induced dimension reduction theorem which
+ * provides a way to construct subsequent residuals that lie in a sequence
+ * of shrinking subspaces.
+ *
+ * @tparam ValueType precision of the elements of the system matrix.
+ *
+ * @ingroup idr
+ * @ingroup solvers
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision>
+class Idr : public EnableLinOp<Idr<ValueType>>,
+            public Preconditionable,
+            public Transposable {
+    friend class EnableLinOp<Idr>;
+    friend class EnablePolymorphicObject<Idr, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using transposed_type = Idr<ValueType>;
+
+    /**
+     * Gets the system operator (matrix) of the linear system.
+     *
+     * @return the system operator (matrix)
+     */
+    std::shared_ptr<const LinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
+
+    std::unique_ptr<LinOp> transpose() const override;
+
+    std::unique_ptr<LinOp> conj_transpose() const override;
+
+    /**
+     * Return true as iterative solvers use the data in x as an initial guess.
+     *
+     * @return true as iterative solvers use the data in x as an initial guess.
+     */
+    bool apply_uses_initial_guess() const override { return true; }
+
+    /**
+     * Gets the stopping criterion factory of the solver.
+     *
+     * @return the stopping criterion factory
+     */
+    std::shared_ptr<const stop::CriterionFactory> get_stop_criterion_factory()
+        const
+    {
+        return stop_criterion_factory_;
+    }
+
+    /**
+     * Sets the stopping criterion of the solver.
+     *
+     * @param other  the new stopping criterion factory
+     */
+    void set_stop_criterion_factory(
+        std::shared_ptr<const stop::CriterionFactory> other)
+    {
+        stop_criterion_factory_ = std::move(other);
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Criterion factories.
+         */
+        std::vector<std::shared_ptr<const stop::CriterionFactory>>
+            GKO_FACTORY_PARAMETER(criteria, nullptr);
```

Comment:
We use `GKO_FACTORY_PARAMETER_VECTOR` or `GKO_FACTORY_PARAMETER_SCALAR` in ginkgo. `GKO_FACTORY_PARAMETER` is only left for keeping interface.
```suggestion
            GKO_FACTORY_PARAMETER_VECTOR(criteria, nullptr);
```
The other parameters should be also changed